server/routerlicious/packages/services-client/src/array.ts

```diff
@@ -30,3 +32,72 @@ export function convertSortedNumberArrayToRanges(numberArray: number[]): number[
 	ranges.push([begin, end]);
 	return ranges;
 }
+
+export function mergeArrays<T, TComp>(arr1: T[], arr2: T[], selector: (item: T) => TComp): T[] {
+	const mergedResult: T[] = [];
+	let index1 = 0;
+	let index2 = 0;
+	while (index1 < arr1.length && index2 < arr2.length) {
+		const value1 = selector(arr1[index1]);
+		const value2 = selector(arr2[index2]);
+		if (value1 <= value2) {
+			mergedResult.push(arr1[index1]);
+			index1++;
+		} else if (value1 > value2) {
+			mergedResult.push(arr2[index2]);
+			index2++;
+		}
+	}
+	while (index1 < arr1.length) {
+		mergedResult.push(arr1[index1]);
+		index1++;
+	}
+	while (index2 < arr2.length) {
+		mergedResult.push(arr2[index2]);
+		index2++;
+	}
+	return mergedResult;
+}
+
+class HeapNode<T> {
+	constructor(
+		public readonly index: number,
+		public readonly row: number,
+		public readonly value: T,
+	) {}
+}
+
+export function mergeKArrays<T>(arrays: T[][], comparator: (a: T, b: T) => number): T[] {
```

Comment:
Not sure if we can use some built-in method?

Reply:
Same, want to achieve the best performance